prow/test/integration/cmd/fakegcsserver/main.go

```diff
@@ -0,0 +1,131 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// fakegcsserver runs the open source GCS emulator from
+// https://github.com/fsouza/fake-gcs-server.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/fsouza/fake-gcs-server/fakestorage"
+	"github.com/sirupsen/logrus"
+
+	configflagutil "k8s.io/test-infra/prow/flagutil/config"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+type options struct {
+	// config is the Prow configuration. We need this to read in GCS
+	// configurations under plank's default_decoration_config_entries field set
+	// in the integration test's Prow configuration, because we have to
+	// initialize (create) these buckets before we upload into them with
+	// initupload.
+	config              configflagutil.ConfigOptions
+	emulatorPort        uint
+	emulatorPublicHost  string
+	emulatorStorageRoot string
+}
+
+func (o *options) validate() error {
+	if o.emulatorPort > 65535 {
+		logrus.Fatal("-emulator-port range must be 0 - 65535")
+	}
+
+	return nil
+}
+
+func flagOptions() *options {
+	o := &options{config: configflagutil.ConfigOptions{ConfigPath: "/etc/config/config.yaml"}}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	fs.UintVar(&o.emulatorPort, "emulator-port", 8888, "Port to use for the GCS emulator.")
+	fs.StringVar(&o.emulatorPublicHost, "emulator-public-host", "fakegcsserver.default:80", "Address of the GCS emulator as seen by other Prow components in the test cluster.")
+	fs.StringVar(&o.emulatorStorageRoot, "emulator-storage-root", "/gcs", "Folder to store GCS buckets and objects.")
+	o.config.AddFlags(fs)
+
+	fs.Parse(os.Args[1:])
+
+	return o
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	o := flagOptions()
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid arguments.")
+	}
+
+	health := pjutil.NewHealth()
+	health.ServeReady()
+
+	defer interrupts.WaitForGracefulShutdown()
+
+	initialObjects, err := getInitialObjects(o)
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not initialize emulator state")
+	}
+
+	logrus.Info("Starting server...")
+
+	server, err := fakestorage.NewServerWithOptions(fakestorage.Options{
+		Scheme:          "http",
+		Host:            "0.0.0.0",
+		Port:            uint16(o.emulatorPort),
+		PublicHost:      o.emulatorPublicHost,
+		Writer:          logrus.New().Writer(),
+		StorageRoot:     o.emulatorStorageRoot,
+		BucketsLocation: "US-CENTRAL1",
+		InitialObjects:  *initialObjects,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	logrus.Infof("Server started at %s", server.URL())
+	logrus.Infof("PublicURL: %s", server.PublicURL())
```

Comment:
is `server.PublicURL()` blocking? if not wouldn't main exit after this line?

Reply:
NewServerWithOptions starts the server. It's a weird design but it is what it is. https://github.com/fsouza/fake-gcs-server/blob/v1.38.3/fakestorage/server.go#L180

Reply:
Oh I see what you mean.

I don't know why but in local tests (before I did the changes I thought you requested) the GCS server keeps running even though those print calls are not blocking. I don't fully understand why this is.

TODO: understand why fake GCS server runs fine even though it appears we are supposed to exit after `main` exits, and add as a comment.